### ISSUE TITLE
Not blocking read_message

### ISSUE THREAD

**Issue by Leonardojrr** [snapview/tungstenite-rs#188]
Hello, I'm starting using tungstenite for one personal project so I'm very noob with this library, be patient please.

I have been wondering if there is a way to read incoming messages without block the thread until a message is received. I want to read and write messages simultaneously but I can't because of this. I have been reading the source code, apparently the method read_message() of the structure WebSocket calls the method read_message() of his field "context" (the field "context" is a WebSocketContext structure). The method has the following code:  [source](https://docs.rs/tungstenite/0.13.0/src/tungstenite/protocol/mod.rs.html#297-315)

Like I see, this method have a loop and is going to wait to the method read_message_frame() to return a message to finally exit the loop. However, there is a method write_pending() that use and apparently method named no_block(). I tried to use this same method no_block() with read_message() but unfortunately it still block the thread until a new message arrive.

Thanks for reading!!!

**Comment by agalakhov** [comment]
Hi,

TL;DR: use `tokio` and `tokio-tungstenite`.

The actually blocking part has nothing to do with WebSockets. It is `TcpStream` and its `read()` call. Tungstenite will be non-blocking if used on top of non-blocking `TcpStream`. You don't have to change anything in Tungstenite, just make your `TcpStream` non-blocking.

Using non-blocking streams require you to do the blocking on your own. All calls to Tungstenite will be non-blocking, that is, they will return immediately. Implementing correct event waiting (blocking) is surprisingly hard. While this could be done with i.e. `mio`, I suggest using higher-level frameworks. The one that fits most needs is `tokio`. Tungstenite has already a Tokio binding called `tokio-tungstenite` and similar ones `async-tungstenite` and `ws_stream_tungstenite`. Other options include `smol`, `may`, `kay` and many web frameworks.

Thank you for using Tungstenite!

**Comment by application-developer-DA** [comment]
That's a common question that we get quite often (especially from beginners in Rust). We've recently added [a "Note" to our "Introduction"](https://github.com/snapview/tungstenite-rs#tungstenite) section to address similar questions and point out to `tokio-tungstenite` (unless you're willing to implement their own runtime to properly manage non-blocking sockets and related logic).

**Comment by stackinspector** [comment]
> Other options include `smol`, `may`, `kay` and many web frameworks.
How can tungstenite support smol? Just by replacing `TcpStream` with `Async<TcpStream>`?

**Comment by daniel-abramov** [comment]
We don't really have a `TcpStream` inside `tungstenite`, apart from the `connect` helper that is meant for the simple use-cases. The rest of the code work with any generic stream that implements `Read + Write`. To support `smol` you would need to wrap the stream into some type that can be passed to `tungstenite`. A good example would be `tokio-tungstenite` and its `compat.rs` layer that ensures that we can use Tokio types with `tungstenite`. You may want to use something similar with `smol` in which case you would probably end up having a thin wrapper crate called `smol-tungstenite` or something like this.

**Comment by stackinspector** [comment]
I found out that the "smol ecosystem" does not exist, it is actually the async-std ecosystem without using the async-std crate itself......
Anyway, async-tungstenite with async-tls feature is a good choice: it's runtime independent, which can be put into any asynchronous TcpStream.
